jdaviz/configs/mosviz/plugins/parsers.py

```diff
@@ -487,16 +489,51 @@ def mos_meta_parser(app, data_obj, ids=None, spectra=False, sp1d=False):
     with app.data_collection.delay_link_manager_update():
 
         # this conditional must mirror the one above
-        if spectra and not sp1d:
+        if data_type == "2D Spectra":
             _add_to_table(app, filters_gratings, "Filter/Grating")
-        elif spectra and sp1d:
-            _add_to_table(app, names, "Identifier")
-        else:
+        elif data_type == "1D Spectra":
             _add_to_table(app, names, "Identifier")
+        elif data_type == "Images":
             _add_to_table(app, ra, "R.A.")
             _add_to_table(app, dec, "Dec.")
 
 
+def _get_source_identifiers(app, data_type, hdus, filepaths=None,
+                            header_keys=['SOURCEID', 'OBJECT']):
+
+    ids = _get_source_identifiers_by_meta(app, data_type, header_keys)
+    # If we fellback for ALL sources, try searching by hdu
+    if set(ids) == set(FALLBACK_NAME):
+        ids = _get_source_identifiers_by_hdu(hdus, filepaths, header_keys)
+    return ids
+
+
+def _get_source_identifiers_by_meta(app, data_type, header_keys=['SOURCEID', 'OBJECT']):
+    
+    src_names = list()
+    # If the user only provided a key that's not already in a container or list, put it in
+    # one for the upcoming loop
+    if not (isinstance(header_keys, Iterable) and not isinstance(header_keys, (str, dict))):
```

Comment:
Can this be shortened?
```suggestion
    if not isinstance(header_keys, (Iterable, str, dict)):
```

Reply:
Careful! The parentheses is important here: the first `not` applies to the whole condition. The crux of this is that `str` is technically iterable... That being said, I don't know why I originally wanted to wrap a list around a dict....

I updated the check to be more explicit/simple: https://github.com/spacetelescope/jdaviz/pull/1851/commits/94ef234a529af86be54f1feb63ab939361985aac